sklearn/metrics/_regression.py

```diff
@@ -1369,9 +1372,10 @@ def _mean_tweedie_deviance(y_true, y_pred, sample_weight, power):
         dev = 2 * (xp.log(y_pred / y_true) + y_true / y_pred - 1)
     else:
         dev = 2 * (
-            xp.pow(y_true, xp.asarray(2 - p)) / ((1 - p) * (2 - p))
-            - y_true * xp.pow(y_pred, xp.asarray(1 - p)) / (1 - p)
-            + xp.pow(y_pred, xp.asarray(2 - p)) / (2 - p)
+            xp.pow(y_true, xp.asarray(2 - p, device=y_true.device))
+            / ((1 - p) * (2 - p))
+            - y_true * xp.pow(y_pred, xp.asarray(1 - p, device=y_true.device)) / (1 - p)
+            + xp.pow(y_pred, xp.asarray(2 - p, device=y_true.device)) / (2 - p)
         )
```

Comment:
Same comments here: for scalar values we should not need to move to the device explicitly, but if we do we should also set the dtype explicitly.

Reply:
Is the above comment applicable i.e. do we need to set the device for scalars and if so should we consider specifying the dtype? Basically is this only a requirement for array-api-strict as I think we weren't really encountering issues with scalars for Pytorch and Cupy?

Reply:
I think we need to explicitly convert scalars to `xp` arrays so that we can specify the device those arrays should be placed on. This is because array-api-strict is more strict about this than other libraries.

The array API standard has been updated to allow scalars for a lot of things, so that you don't have to explicitly create arrays. But it will take some time before that change is formalised and arrives in all the libraries/we can rely on it.

With changing to use `lgamma` we have less `xp.asarray` stuff, which is nice.

Reply:
Yes I think that makes sense as array-api-strict has been the main limitation with these kinds of situations. In any case do you think we should specify the `dtype` as well in the above `xp.asarray` statements?

Reply:
I don't know :-/ Naively it seems like "choose what ever dtype you want" is fine here? In particular because at the end of the function we convert to a Python `float` anyway. But maybe it matters to numerical precision??

Reply:
Sorry, I started a new related thread by trying to suggest a change on those particular lines:

https://github.com/scikit-learn/scikit-learn/pull/30090#discussion_r1979839022